deploy-tool/dns/dns.go

```diff
@@ -0,0 +1,35 @@
+package dns
+
+import (
+	"fmt"
+	"github.com/caddyserver/certmagic"
+	"github.com/libdns/cloudflare"
+	"os"
+	"strings"
+)
+
+// GetACMEDNSProvider configures an ACMEDNSProvider value to be used in cert generation
+func GetACMEDNSProvider() (certmagic.ACMEDNSProvider, error) {
+	switch strings.ToLower(getDNSProvider()) {
+	case "cloudflare":
+		provider := &cloudflare.Provider{
+			APIToken: getAPIToken(),
+		}
+
+		return provider, nil
+	}
+
+	return nil, fmt.Errorf("no valid provider selected")
+}
+
+func getDNSProvider() string {
+	dnsProvider := os.Getenv("D_DNS_PROVIDER")
+
+	return dnsProvider
+}
+
+func getAPIToken() string {
+	apiToken := os.Getenv("D_API_TOKEN")
```

Comment:
We gotta talk about how to manage this API token.

Reply:
Why would we need an API token for the LE ACME endpoint?

Reply:
There is a notion of an account and a user in the ACME API: https://tools.ietf.org/html/rfc8555#section-7.3

I think maybe certmagic might be abstracting that away from you, and maybe what they're doing is a brand new user for each certificate request. We should check what's going on, and whether it'd be better to manage a persistent ACME credential (maybe ask the SRE and boulder teams for opinions).

Reply:
Oh you mean the user private key. Yeah since when I wrote this we were planning to only run this on our local machines I assumed we're just going to handle the key.

I think if the intention of this (or another tool that reuses some of the logic) to run automatically in k8s the proper way to do this would be to use the k8s secret storage or just mount it as a volume for the container to see.